### ISSUE TITLE
Use multiple Workspaces to support path mapping across multiple Kythe projects [phab:D1912]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2355]
_Created by **djrenren** at 2017-08-28 23:48:30_: 

Previously, KLS used a single "PathConfig" object to control path mapping.
This made it impossible to open multiple kythe projects at once. Now the server uses
multiple `Workspace`s which perform the same task as a singular PathConfig (though could
in the future adopt more responsibility such as Workspace-local xref services). When a new
file is opened, the server will give each existing `Workspace` a chance to discover the file
(in the case of SettingsWorkspace it just looks for a path prefix). If all the workspaces file,
it falls back to a `WorkspaceProvider` function (supplied at instantiation) to produce a new
workspace.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

Why does this need to be given a name at all? There's no abstraction here—it's just a callback. Please don't make the user click to another file just to see a type declaration this simple.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

Please don't name functions after nouns.

This could more simply be

    func NewSettingsWorkspace(…) (Workspace, error) {

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

Again as below, please don't name functions after nouns.

    // Construct a workspace from a document URI
    newWorkspace func(lsp.DocumentURI) (Workspace, error)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

`Path` or `RelativePath` maybe—something to indicate that it's a path.

    Path string // the path of the file relative to the workspace root

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

`_, err := os.Stat(filepath.Join(path, settingsFile)`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

`p` → `path`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 14:33:52_:

I feel like you just implemented this logic (walking up a directory chain) somewhere else nearby. Maybe time to pull that out?

```
// FindEnclosing returns the nearest enclosing directory d of dir for which match(d)
// returns true. If no such directory is found, FindEnclosing returns os.ErrNotExist.
// Any other error is returned immediately.
func FindEnclosing(dir string, match func(string) bool) (string, error) {
	cur := filepath.Clean(dir)
	for {
		fs, err := os.Stat(cur)
		if err != nil {
			return "", err
		} else if !fs.IsDir() {
			return "", fmt.Errorf("not a directory: %q", cur)
		} else if match(cur) {
			return cur, nil
		} else if cur == "/" || cur == "." {
			break
		}
		cur = filepath.Dir(cur)
	}
	return "", os.ErrNotExist
}
```

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-29 18:41:54_:

- Generalize findRoot as FindEnclosing
- Remove WorkspaceProvider
- Rename LocalFile.Relative to LocalFile.RelativePath

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 19:28:50_:

Are there other things we construct SettingsWorkspaces from? If not the "FromURI" can probably go.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 19:28:50_:

This is merely a string, 

   if gl.RelativePath != rel { …

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-08-29 19:28:50_:

P'raps

    fi, err := os.Stat(filepath.Join(d, settingsFile)
    return err == nil && fi.IsRegular()

if you want to ensure it's "sane".

**Comment by kythe** [comment]
_**djrenren** wrote at 2017-08-29 21:07:34_:

  - Check FileMode in findSettings
  - Remove extraneous deepEqual

**Comment by kythe** [comment]
_**djrenren** commented inline (context lost during migration) at 2017-08-29 21:37:33_:

Yeah, the normal constructor just takes a Settings object. (see Line:89).

This is used in tests where we don't want to mock out filesystem access and in the implementation of alternative Workspace constructors where we have alternate schemes of discovery settings but want the same behavior for path mapping.

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2017-08-29 21:39:57_

**Comment by kythe** [comment]
_**djrenren** closed via commit 0185288b30143d298ef0979d87751f54c8db4a68 at 2017-08-29 22:01:30_
